api/src/Entity/CampCollaboration.php

```diff
@@ -158,6 +159,7 @@ class CampCollaboration extends BaseEntity implements BelongsToCampInterface {
      * @ORM\ManyToOne(targetEntity="Camp", inversedBy="collaborations")
      * @ORM\JoinColumn(nullable=false, onDelete="cascade")
      */
+    #[AssertHasAtLeastOneManager]
```

Comment:
Ich finde es etwas unintuitiv, den Validator auf dem `camp` property zu setzen. Gerade dieser Wert kann ja gar nicht überschrieben werden und damit im eigentlich Sinn auch nicht validiert werden. The camp collaborations kriegst du im Validator auch via `$this->context->getObject()->camp->collaborations`

Die eigentlichen Werte die wirklich ändern und validiert werden mussten sind ja eigentlich `status` und `role`.

Wie wäre es mit sowas:

```
#[Assert\NoChangeIfLastManager]
public string $status = self::STATUS_INVITED;

#[Assert\NoChangeIfLastManager]
public string $role;
```

oder sowas:
```
#[Assert\EqualToIfLastManager(value: self::STATUS_ESTABLISHED)]
public string $status = self::STATUS_INVITED;

#[Assert\EqualToIfLastManager(value: self::ROLE_MANAGER)]
public string $role;
```

Reply:
Stimmt, oder eine Klassen-Validierung?

Reply:
Daran hab ich auch gedacht. Kann API Platform das? Oder müsste  wir manuell einbauen über den Data Persister?

Reply:
Symfony kann das: https://symfony.com/doc/current/validation/custom_constraint.html#class-constraint-validator

Reply:
Mir sind 5 Varianten eingefallen, um das umzusetzen:

![grafik](https://user-images.githubusercontent.com/1506818/148675662-d5cd3435-759f-4873-b0bc-93b738641364.png)


[URL](https://www.plantuml.com/plantuml/uml/bP9BReGm38RtEON5fb8hYgQFJLOZSOVk3MHc8YKnIfpPLBrx0uC2251LLZRc_tu_E0T6pz4QyNHZxeEigjwW6zwyjcIhrc08PQVnwj4ACQPGe-tXMq1wNg0bOx0XZwp91V4Zbb_Lo-8iIPKfJFOSmvHuCce0ZFpYL41PGl5SR4b2E685HbqOwBAjIvajlj1hTApW9E6EHdUFLZ0UovJh1lnaulNrTkGoTvzjtcIAiytWnEnr4rc53FwJEou2_ejzS1jY9hiB1Za3toNKuvdkA-n6ixU51URZ6BDp92dNVyjwni_LZ1WFD7dSSugYDRy0)

1. Variante: Validation auf CampCollaboration::camp
Ist in der nähe wo der CampCollaboration wo der Patch passiert, und sagt das aus, was geprüft werden soll:
Das Camp soll gültig sein, momentan vor allem indem es einen Manager hat.
Man muss nur die implementation der Validation ändern, wenn sich an der Definition für "Collaborator ist Manager" etwas ändert
Nachteil: wenn man role oder status ändert, ist plötzlich die camp Property ungültig.
Für ein api-select ist es dann schwierig, z.b. Fehlermeldungen zuzuordnen.
Falls die Anzahl Manager noch an einem anderen Ort geändert werden kann, müsste man die Validation dort auch einbauen.

2. Variante: Validation auf Attributen
Ist wirklich auf den Attributen, auf denen der Patch passiert.
api-select kann die Fehlermeldung zuordnen.
Falls sich die Definition für "Collaborator ist Manager" ändert, muss man evtl Validations auf Attribute hinzufügen, ändern oder löschen.
Der Wert, den man auf das Attribut schickt, ist nicht wirklich ungültig, sondern das camp wird sodurch ungültig. Ist eher imperativ (Du darfst nicht ändern) als deklarativ (Camp muss einen Manager haben).

3. Variante: Class Validation auf CampCollaboration
Ist in der nähe wo der CampCollaboration wo der Patch passiert, und sagt das aus, was geprüft werden soll:
Das Camp soll gültig sein, momentan vor allem indem es einen Manager hat.
Man muss nur die implementation der Validation ändern, wenn sich an der Definition für "Collaborator ist Manager" etwas ändert
Nachteil: wenn man role oder status ändert, ist plötzlich die camp Property ungültig.
Für ein api-select ist es dann schwierig, z.b. Fehlermeldungen zuzuordnen.
Diese Variante fühlt sich für mich komisch an.

4. Variante: Class Validation auf Camp
Braucht zusätzlich ein [#Valid] auf CampCollaboration::Camp
Ist weiter weg von der CampCollaboration
Für ein api-select ist es dann schwierig, z.b. Fehlermeldungen zuzuordnen.
Fühlt sich aus objektorientierter Perspektive richtig an.
Wir haben uns aber schonmal dagegen entschieden, evtl sind sind die Gründe dafür verschwunden. (https://github.com/ecamp/ecamp3/pull/1280#discussion_r615440992)

5. Variante: Validation auf Camp::collaborations
Braucht zusätzlich ein [#Valid] auf CampCollaboration::Camp
Ist weiter weg von der CampCollaboration
Für ein api-select ist es dann schwierig, z.b. Fehlermeldungen zuzuordnen.
Fühlt sich aus objektorientierter Perspektive richtig an.
Ist so wie wir sicherstellen, dass ein Camp mindestens eine Periode hat.

Ich persönlich hätte Variante 4 oder 5 verwendet, aber weil wir das schonmal nicht wollten,
habe ich jetzt die aus meiner sicht einfache variante 1 gewählt.
Was meint ihr?

Reply:
Mir gefällt Variante 5 auch am besten. Um die beiden Argumente von @pmattmann in #1280 zu addressieren (die ich damals unterstützt hatte):

> Wenn jemand die Methode CampService::validateCamp ergänzt, müsste er hier das Fehler-Mapping anpassen. Das geht leicht vergessen.

Ich stehe weiterhin dazu, dass ich das Mapping von Validation Errors vom Camp nach `CampCollaboration::owner` awkward finde, und dass das mit der Zeit vergessen gehen wird. Ich würde mir die unabgeänderten Fehlermeldungen von API Platform vorstellen:
```json
{
  "propertyPath":"camp.collaborations",
  "message":"At least one manager must remain in the camp.",
  "code":"c1051bb4-d103-4f74-8988-acbcafc7fdc3"
}
```
Das api-select kann dann bei jeder Validation Message anhand des `propertyPath`s entscheiden, diese Message direkt beim Select zu rendern, oder sie weiterzugeben an einen allgemeinen `<server-error>` Bereich (oder Toast / Snackbar etc.) der für das ganze Formular gilt. So einen Bereich brauchen wir sowieso, um dem User allgemeine 500er etc. zu melden. Der `propertyPath` kann zudem auch übersetzt werden: "Lager &rarr; Leitungsteam: Mindestens ein Manager muss im Lager bleiben". Und der mit dem Autosave und roter Markierung des api-selects sollte dem User ja dann klar sein, welche seiner Eingaben jetzt diesen Fehler verursacht hat.

> Falls das Camp aus anderen Gründen nicht Valid ist, kann niemand mehr das Camp verlassen. Das ist eine seltsame und unnötige Abhängigkeit.

Das gilt tatsächlich weiterhin. Aber vielleicht liesse sich das mit ValidationGroups lösen? So was ala
```php
class Camp {
  #[AssertAtLeastOneManager(groups: ['campCollaboration:Default'])]
  public Collection $collaborations;
}
```
Das wird so in dieser Form noch nicht funktionieren weil die standard-Validierungsgruppe immer `Default` ist, aber vielleicht gibts eine Lösung in dieser Ecke?
Für spätere Referenz hier ein Issue bei Symfony wo es leider nicht danach klingt als könnte man einfach eine andere Validation Group für eine nested Entity setzen: https://github.com/symfony/symfony/issues/30094

Reply:
In erster Linie hätte ich mich für **Variante 2 oder 3** ausgesprochen. Es geht um Änderungen auf dem Endpoint/Entity CampCollaboration und gewisse Änderungen/Zustände sind (im Kontext aller anderen CampCollaborations) gültig oder nicht gültig.

**Variante 1** finde ich wie bereits oben erwähnt nicht sehr intuitiv da:
- das camp property nicht (bzw. nie) geändert wird
- mit dem gleichen Argument fast alle Validations ans camp gehängt werden können (das Camp is ungültig wenn es keine Periode hat, das Camp ist (evt.) ungültig wenn es keine Kategorie hat, etc.).

Gleiche Argumentation für **Variante 4**.

**Variante 5:** Hätte ich jetzt auf Anhieb nicht daran gedacht. Liest sich grundsätzlich aber relativ intuitiv. Kann aber auch verwirrend sein, wenn jemand auf CampCollaboration die Logik dazu sucht aber nichts dazu findet.
Ein paar Kommentare:
- Technisch funktioniert das einwandfrei, sowohl wenn ich campCollaborations embedded via Camp aktualisieren würde als auch wenn ich direkt auf dem CampCollaboration endpoint operiere?
- Bei Periods: Dort ist es aktuell eben genau nicht vollständig implementiert. Beim Erstellen springt die Validierung zwar an, ich kann die letzte Periode aber danach wieder löschen (siehe https://github.com/ecamp/ecamp3/issues/1278#issuecomment-1008030337).
- Würde das diesen Fall (Löschvorgang) auch unterstützen? Bei CampCollaboration ist das ja nicht möglich, aber bei Period müssten wir genau das implementieren wollen.

Reply:
Das sollte möglich sein, da kommen wir wahrscheinlich auf ein ähnliches Problem wie "Löschen von CampCollaboration is nur zulässig, wenn der status INACTIVE ist.

Technisch funktioniert das einwandfrei, sowohl wenn ich campCollaborations embedded via Camp aktualisieren würde als auch wenn ich direkt auf dem CampCollaboration endpoint operiere?

-> Kann ich nicht sagen, da embedded patch auf einer collection glaube ich momentan nicht funktioniert.

"
* Bei Periods: Dort ist es aktuell eben genau nicht vollständig implementiert. Beim Erstellen springt die Validierung zwar an, ich kann die letzte Periode aber danach wieder löschen (siehe
Letzte Periode nicht löschbar #1278 (comment)).

* Würde das diesen Fall (Löschvorgang) auch unterstützen? Bei CampCollaboration ist das ja nicht möglich, aber bei Period müssten wir genau das implementieren wollen.
"

Api Platform lässt keine validations laufen bei einer DELETE Operation. Das muss man manuell ankicken, so wie es im CampCollaborationDataPerister gemacht ist.

Reply:
Done with https://github.com/ecamp/ecamp3/pull/2363/commits/056243c8ec63b11160e9a97ef18ac450f7dfa697